### PR TITLE
[`pylint`] Implement `too-many-ancestors` (`PLR0901`)

### PR DESCRIPTION
**Pull request by Peiffap**
## Summary

This PR adds [`pylint` rule `R0901`](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-ancestors.html), which restricts the maximum number of parent classes a class can have (to a default value of 7).

## Test Plan

This PR was tested using the workflow discussed [here](https://docs.astral.sh/ruff/contributing/).

Running tests with `cargo test` gives no failures.

## Related Issues
This is part of the greater effort of #970 (and if merged, should be checkmarked as "done" there).

### CODE REVIEW

**Comment by Peiffap** [comment]
Currently, the error snapshot shows that the error message covers the entirety of the class, including the class body. Would it be desirable/easily feasible to only have it cover the "header" instead?

File: crates/ruff_workspace/src/configuration.rs [hunk 1]
```diff
@@ -1502,7 +1502,7 @@ mod tests {
         Rule::ReimplementedStarmap,
         Rule::SliceCopy,
         Rule::TooManyPublicMethods,
-        Rule::TooManyPublicMethods,
+        Rule::TooManyAncestors,
```
Comment (Peiffap) [inline crates/ruff_workspace/src/configuration.rs:1505]: I'm assuming this wasn't meant to be a duplicate, but for visibility I'm adding this comment.

File: crates/ruff_linter/src/rules/pylint/snapshots/ruff_linter__rules__pylint__tests__too_many_ancestors.snap
Comment (Peiffap) [inline crates/ruff_linter/src/rules/pylint/snapshots/ruff_linter__rules__pylint__tests__too_many_ancestors.snap:1]: Currently, the error message covers the entirety of the class, including the class body. Would it be desirable/easily feasible to only have it cover the "header" instead?

**Comment by MichaReiser** [comment]
Thanks for contributing this rule. 

I worry that merging the rule as is today will be confusing to users and give them a wrong sense of "safety" where they think the rule would catch all inheritance chains with more than `x` parent classes. However, that's not the case because Ruff doesn't support multifile analysis, meaning Ruff can only reason about the ancestor chain in a single file. 

I think it's more likely that such deep hierarchy chains are split across many files, meaning the rule will miss most real violations. 

I'll close this PR for now but I think we can add this rule once multifile analysis is supported.

**Comment by Peiffap** [comment]
@MichaReiser Thanks for taking the time to look over this!

If I'm allowed to push back a little on this... I don't understand your comment at all :(. This PR (and to my understanding, the `pylint` rule associated to it) has nothing to do with the _depth_ of the inheritance chain (which would definitely be an interesting rule to implement once multifile analysis is possible!), but about the _width_ (if you can call it that) of the inheritance, i.e. the number of _direct_ parent classes of a given class.
I'm by no means an expert on object-oriented Python, but to my knowledge this is information that is solely contained in a single file, where the class is defined.

**Comment by MichaReiser** [comment]
I'm sorry that I failed to communicate the decision clearly or may even be mistaken. 

I re-read the rule documentation and I think there's some confusion because the documentation sometimes uses the term parents and sometimes ancestors:

* `parents`: The direct parents of a class. This maps to `class_def.bases`
* `ancestors`: The transitive inheritance chain of a class. This requires traversing the entire inheritance chain. 

I took a look at the Pylint implementation, and my understanding is that it asserts that `ancestors` is less than the configured minimum, whereas this implementation implements `parents`. 

https://github.com/pylint-dev/pylint/blob/087d77adcd8455dbbbf6b43631d93abb2d1c5d67/pylint/checkers/design_analysis.py#L245-L279

We'll need multifile analysis if we want to implement Pylint's behavior (assuming I understand it correctly)

**Comment by Peiffap** [comment]
Thanks for giving some more detail! Apologies if I came across as overly defensive.

I agree with you that the documentation is a little confusing and mixes terms, and I wasn't even aware myself that "parent" and "ancestor" refer to different things!
I'm not familiar enough with the ruff codebase to know whether multifile analysis is required to check all ancestors, so I'll take your word for it.
How would you feel about slightly refactoring this PR to implement a new "`too-many-parents`" rule which clearly explains the difference, and adding a comment to #970 specifying that `R0901` shouldn't be worked on until multifile analysis is available?

**Comment by MichaReiser** [comment]
> Apologies if I came across as overly defensive.

No worries. It was good that you asked for clarification because I honestly wasn't aware of what specific behavior PyLInt implements (I just assumed ancestor without checking). 

> I'm not familiar enough with the ruff codebase to know whether multifile analysis is required to check all ancestors, so I'll take your word for it.

We could check all ancestors already today, but only if all classes are defined in the same file. That's unlikely to be the case for many classes and significantly reduces the added value of the rule.

> How would you feel about slightly refactoring this PR to implement a new "too-many-parents" rule which clearly explains the difference, 

I prefer to wait until Ruff supports multifile-analysis to avoid that we have to deprecate this rule just in a few months because it overlaps with an other rule. I also think that only checking direct parents isn't that useful, at least not in the spirit the rule is defined today (you could argue that inheriting from more than one base is considered bad). 

> and adding a comment to https://github.com/astral-sh/ruff/issues/970 specifying that R0901 shouldn't be worked on until multifile analysis is available?

Yeah, let's do that

**Comment by Peiffap** [comment]
That's all reasonable; thanks for getting back to me so swiftly! I'll add the comment as discussed.

**Pull request closed without merging**